pkg/operator/externaletcdsupportcontroller/external_etcd_support_controller.go

```diff
@@ -0,0 +1,192 @@
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/resource/resourceapply"
+	"github.com/openshift/library-go/pkg/operator/resource/resourceread"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	coreclientv1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+)
+
+type ExternalEtcdEnablerController struct {
+	operatorClient v1helpers.StaticPodOperatorClient
+
+	targetImagePullSpec   string
+	operatorImagePullSpec string
+	envVarGetter          etcdenvvar.EnvVar
+
+	kubeClient kubernetes.Interface
+
+	enqueueFn func()
+}
+
+func NewExternalEtcdEnablerController(
+	operatorClient v1helpers.StaticPodOperatorClient,
+	targetImagePullSpec, operatorImagePullSpec string,
+	envVarGetter etcdenvvar.EnvVar,
+	kubeInformersForOpenshiftEtcdNamespace informers.SharedInformerFactory,
+	kubeInformersForNamespaces v1helpers.KubeInformersForNamespaces,
+	infrastructureInformer configv1informers.InfrastructureInformer,
+	networkInformer configv1informers.NetworkInformer,
+	masterNodeInformer cache.SharedIndexInformer,
+	kubeClient kubernetes.Interface,
+	eventRecorder events.Recorder) factory.Controller {
+
+	c := &ExternalEtcdEnablerController{
+		operatorClient:        operatorClient,
+		targetImagePullSpec:   targetImagePullSpec,
+		operatorImagePullSpec: operatorImagePullSpec,
+		envVarGetter:          envVarGetter,
+		kubeClient:            kubeClient,
+	}
+	syncCtx := factory.NewSyncContext("ExternalEtcdSupportController", eventRecorder.WithComponentSuffix("external-etcd-support-controller"))
+	c.enqueueFn = func() {
+		syncCtx.Queue().Add(syncCtx.QueueKey())
+	}
+	envVarGetter.AddListener(c)
+	syncer := health.NewDefaultCheckingSyncWrapper(c.sync)
+
+	return factory.New().
+		WithSyncContext(syncCtx).
+		ResyncEvery(time.Minute).
+		WithSync(syncer.Sync).
+		WithInformers(
+			operatorClient.Informer(),
+			kubeInformersForNamespaces.InformersFor(operatorclient.TargetNamespace).Core().V1().Endpoints().Informer(),
+			kubeInformersForOpenshiftEtcdNamespace.Core().V1().ConfigMaps().Informer(),
+			kubeInformersForOpenshiftEtcdNamespace.Core().V1().Secrets().Informer(),
+			masterNodeInformer,
+			infrastructureInformer.Informer(),
+			networkInformer.Informer(),
+		).ToController("ExternalEtcdController", eventRecorder.WithComponentSuffix("external-etcd-controller"))
+}
+
+func (c *ExternalEtcdEnablerController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+
+	operatorSpec, _, _, err := c.operatorClient.GetStaticPodOperatorState()
+	if err != nil {
+		return err
+	}
+
+	envVars := c.envVarGetter.GetEnvVars()
+	if len(envVars) == 0 {
+		// note this will not degrade the controller, that can happen during CEO restarts often due to cold informer caches (expected)
```

Comment:
precisely why you should use `AddListener` and only call `GetEnvVars` after you got a callback

Reply:
then I wonder why you need that length check: there must be something in that slice after that callback. Or is there an informer sync missing on the controller?

Reply:
honestly, I took that check from the other controllers, assuming it was useful :grin: